File: scraper.js
const puppeteer = require('puppeteer');
const cheerio = require('cheerio');
const fs = require('fs'); // Include the file system module

const url = 'https://mason360.gmu.edu/events';

(async () => {
  let browser;
  try {
    // Launch the headless browser and ensure the variable is in the correct scope
    browser = await puppeteer.launch();
  
    // Open a new page
    const page = await browser.newPage();
  
    // Navigate to the URL
    await page.goto(url, { waitUntil: 'networkidle0' });
  
    // Use the page content in cheerio
    const content = await page.content();
    const $ = cheerio.load(content);
    const events = [];

    // Select the elements and get the data
    $('div.media-body > h3.media-heading.header-cg--h4').each(function () {
      const title = $(this).text().trim();
      events.push(title);
    });

    // Save the events to a file
    fs.writeFileSync('events.txt', events.join('\n'));

    console.log('Events have been saved to events.txt');
  } catch (error) {
    console.error('An error occurred:', error);
  } finally {
    // Ensure the browser gets closed even if an error occurs
    if (browser) {
      await browser.close();
    }
  }
})();

File: scraper.py
from requests_html import HTMLSession
from bs4 import BeautifulSoup


url = 'https://mason360.gmu.edu/events'
base_url = 'https://mason360.gmu.edu'

s = HTMLSession()
r = s.get(url)

r.html.render(sleep=2)  

h2_elements = r.html.find('h2.header-cg--h4', first=False)

skip_string = "[eventName] [eventBadges] IF[:hybrid]='hybrid'--> Hybrid"
h3_elements = r.html.find('div.media-body > h3.media-heading.header-cg--h4', first=False)

hrefs = []

for h3 in h3_elements:
    # Find <a> tags within the h3 element
    a_tags = h3.find('a', first=False)
    # Extract the href attribute from each <a> tag
    for a in a_tags:
        href = a.attrs.get('href')
        if href and "rsvp" in href:
            hrefs.append(base_url + href)

with open('events_combined.txt', 'w', encoding='utf-8') as file:
    for newUrl in hrefs:
        new_response = s.get(newUrl)
        new_response.html.render(sleep=1)
        soup = BeautifulSoup(new_response.content, 'html.parser')
        target_divs = soup.find_all('div', class_='col-md-4_5')

        # Write the event URL and title to the file
        title = new_response.html.find('title', first=True).text if new_response.html.find('title') else 'No Title Found'
        file.write(f"{newUrl}\n{title}\n")

        # Iterate through each div and find <p> tags
        for div in target_divs:
            p_tags = div.find_all('p')
            for p in p_tags:
                dwrite = p.get_text(strip=True)
                file.write(dwrite + "\n")  

        # Add a new line to separate events
        file.write("\n")



